File: cmd/unread.go
package cmd

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/mturley/agent-handler/db"
	"github.com/spf13/cobra"
)

var unreadCmd = &cobra.Command{
	Use:   "unread",
	Short: "List unread events for a session",
	RunE:  runUnread,
}

func init() {
	unreadCmd.GroupID = "agent"
	rootCmd.AddCommand(unreadCmd)
	unreadCmd.Flags().String("session-id", "", "session ID (defaults to session from PID cache)")
	unreadCmd.Flags().Bool("ack", false, "acknowledge events after reading")
	unreadCmd.Flags().Bool("agent-only", false, "with --ack, advance only the agent cursor (not human cursor)")
	unreadCmd.Flags().Bool("count", false, "only print the unread count")
}

func runUnread(cmd *cobra.Command, args []string) error {
	ack, _ := cmd.Flags().GetBool("ack")

	var d *db.DB
	var err error
	if ack {
		d, err = openDB()
	} else {
		d, err = openReadOnlyDB()
	}
	if err != nil {
		return fmt.Errorf("failed to open database: %w", err)
	}
	defer d.Close()

	sessionID, err := resolveSessionID(cmd)
	if err != nil {
		return fmt.Errorf("could not determine session: %w", err)
	}

	countOnly, _ := cmd.Flags().GetBool("count")

	if countOnly {
		count, _, err := d.UnreadCountForSession(sessionID)
		if err != nil {
			return fmt.Errorf("failed to count unread events: %w", err)
		}
		fmt.Println(count)
		return nil
	}

	events, err := d.UnreadForSession(sessionID)
	if err != nil {
		return fmt.Errorf("failed to query unread events: %w", err)
	}

	if ack && len(events) > 0 {
		agentOnly, _ := cmd.Flags().GetBool("agent-only")
		ts := time.Now().UTC().Format(time.RFC3339)
		if agentOnly {
			d.AdvanceCursor(sessionID, ts)
		} else {
			d.AdvanceBothCursors(sessionID, ts)
		}
	}

	if jsonOutput {
		data, err := json.MarshalIndent(events, "", "  ")
		if err != nil {
			return err
		}
		fmt.Println(string(data))
	} else {
		if len(events) == 0 {
			fmt.Println("No unread events")
			return nil
		}

		fmt.Printf("Unread events (%d):\n\n", len(events))
		for _, e := range events {
			author := "-"
			if e.Author != nil {
				author = *e.Author
			}
			fmt.Printf("  [%s] %s\n", e.Type, e.Title)
			fmt.Printf("  Author: %s | Time: %s\n", author, e.TS)
			if e.Body != nil && *e.Body != "" {
				fmt.Printf("  %s\n", *e.Body)
			}
			fmt.Println()
		}
	}

	return nil
}

File: src/Handler/Cmd/Unread.hs
-- | Port of cmd/unread.go: list unread events for a session.
module Handler.Cmd.Unread (unreadCommand) where

import Control.Monad (unless, when)
import Data.Aeson (toJSON)
import Data.Maybe (fromMaybe, isJust)
import Data.Text (Text)
import qualified Data.Text as T
import Options.Applicative

import Handler.Cli.Common
import Handler.Db (close)
import Handler.Db.Cursors (advanceBothCursors, advanceCursor)
import Handler.Db.Events
  ( Event(..), eventToJson, globalUnreadCountForSession, globalUnreadForSession
  , unreadCountForSession, unreadForSession )
import Handler.Util (nowIso)

data UnreadOpts = UnreadOpts
  { uSessionId :: Maybe Text
  , uAck       :: Bool
  , uAgentOnly :: Bool
  , uCount     :: Bool
  , uGlobal    :: Bool
  }

unreadCommand :: Mod CommandFields NamedCommand
unreadCommand = mkCommand "unread" "List unread events for a session" (runUnread <$> opts)
  where
    opts = UnreadOpts
      <$> sessionIdOption
      <*> switch (long "ack" <> help "acknowledge events after reading")
      <*> switch (long "agent-only" <> help "with --ack, advance only the agent cursor (not human cursor)")
      <*> switch (long "count" <> help "only print the unread count")
      <*> switch (long "global" <> help "show all events since cursor, not just those targeted at this session (for handler sessions)")

runUnread :: UnreadOpts -> Ctx -> IO ()
runUnread o ctx = do
  db <- if o.uAck then openDb else openReadOnlyDb
  sessionId <- resolveSessionIdOpt o.uSessionId

  if o.uCount
    then do
      (count, _) <- if o.uGlobal
        then globalUnreadCountForSession db sessionId
        else unreadCountForSession db sessionId
      putTextLn (T.pack (show count))
    else do
      allEvents <- if o.uGlobal
        then globalUnreadForSession db sessionId
        else unreadForSession db sessionId

      -- In global mode, filter out events originated by this session
      let events = if o.uGlobal
            then [e | e <- allEvents, e.sessionId /= Just sessionId]
            else allEvents

      when (o.uAck && not (null events)) $ do
        ts <- nowIso
        if o.uAgentOnly
          then advanceCursor db sessionId ts
          else advanceBothCursors db sessionId ts

      if ctx.jsonOutput
        then printJson (toJSON (map eventToJson events))
        else if null events
          then putTextLn "No unread events"
          else do
            printfT ["Unread events (", T.pack (show (length events)), "):\n"]
            mapM_ printEvent events
  close db
  where
    printEvent e = do
      printfT ["  [", e.eventType, "] ", e.title]
      printfT ["  Author: ", fromMaybe "-" e.author, " | Time: ", e.ts]
      unless (not (isJust e.body) || e.body == Just "") $
        printfT ["  ", fromMaybe "" e.body]
      putTextLn ""
